tools/packaging/kata-deploy/local-build/cache-kata-deploy-binaries.sh

```diff
@@ -0,0 +1,280 @@
+	local guest_image_last_commit="$(get_last_modification "${repo_root_dir}/tools/packaging/guest-image")"
+	local agent_last_commit="$(get_last_modification "${repo_root_dir}/src/agent")"
+	local libs_last_commit="$(get_last_modification "${repo_root_dir}/src/libs")"
+	local gperf_version="$(get_from_kata_deps "externals.gperf.version")"
+	local libseccomp_version="$(get_from_kata_deps "externals.libseccomp.version")"
+	local rust_version="$(get_from_kata_deps "languages.rust.meta.newest-version")"
+	local rootfs_tarball_name="kata-static-rootfs-${ROOTFS_FLAVOUR}.tar.xz"
+	local current_rootfs_version="${osbuilder_last_commit}-${guest_image_last_commit}-${agent_last_commit}-${libs_last_commit}-${gperf_version}-${libseccomp_version}-${rust_version}-${ROOTFS_FLAVOUR}"
+	create_cache_asset "${rootfs_tarball_name}" "${current_rootfs_version}" ""
+}
+
+cache_shim_v2_artifacts() {
+	local shim_v2_tarball_name="kata-static-shim-v2.tar.xz"
+	local shim_v2_last_commit="$(get_last_modification "${repo_root_dir}/src/runtime")"
+	local protocols_last_commit="$(get_last_modification "${repo_root_dir}/src/libs/protocols")"
+	local runtime_rs_last_commit="$(get_last_modification "${repo_root_dir}/src/runtime-rs")"
+	local golang_version="$(get_from_kata_deps "languages.golang.meta.newest-version")"
+	local rust_version="$(get_from_kata_deps "languages.rust.meta.newest-version")"
+	local current_shim_v2_version="${shim_v2_last_commit}-${protocols_last_commit}-${runtime_rs_last_commit}-${golang_version}-${rust_version}"
+	local current_shim_v2_image="$(get_shim_v2_image_name)"
+	create_cache_asset "${shim_v2_tarball_name}" "${current_shim_v2_version}" "${current_shim_v2_image}"
+}
+
+cache_virtiofsd_artifacts() {
+	local virtiofsd_tarball_name="kata-static-virtiofsd.tar.xz"
+	local current_virtiofsd_version="$(get_from_kata_deps "externals.virtiofsd.version")-$(get_from_kata_deps "externals.virtiofsd.toolchain")"
+	local current_virtiofsd_image="$(get_virtiofsd_image_name)"
+	create_cache_asset "${virtiofsd_tarball_name}" "${current_virtiofsd_version}" "${current_virtiofsd_image}"
+}
+
+create_cache_asset() {
+	local component_name="${1}"
+	local component_version="${2}"
+	local component_image="${3}"
+
+	local build_dir="${repo_root_dir}/tools/packaging/kata-deploy/local-build/build"
+	local cache_dir="${script_dir}/cache"
+
+	sudo chown -R "${USER}:${USER}" "${build_dir}/${component_name}"
+
+	mkdir -p ${cache_dir}/${cache_target}
+	cp "${build_dir}/${component_name}" "${cache_dir}/${cache_target}/${component_name}"
+	sha256sum "${build_dir}/${component_name}" > "${cache_dir}/${cache_target}/sha256sum"
+	cat "${cache_dir}/${cache_target}/sha256sum"
+	echo "${component_version}" > "${cache_dir}/${cache_target}/latest"
+	cat "${cache_dir}/${cache_target}/latest"
+	echo "${component_image}" > "${cache_dir}/${cache_target}/latest_image"
+	cat "${cache_dir}/${cache_target}/latest_image"
+
+	docker build -t ${REGISTRY}:${cache_target}-${BRANCH}-${ARCH} \
```

Comment:
I'm going to assume that you've ruled it out for good reason before, but in case not (and to let me understand the reason as it might help with caching in other projects): What about using the commit and/or sha256sum/component version as the tag for the artifact's container image (maybe in addition to the current approach). Then if different branches have the same images (which I appreciate it likely to be an uncommon scenario), or if we have to revert back to an older version of builds in same branches (e.g. stable), they could just be pulled directly without only keeping the latest version. Or is the concern about ending up with too many images then?

A more concrete scenario that I'm thinking of is that in the cloud-api-adaptor we pull in the kata-agent binary (which I know isn't currently cached). It's likely that we don't want to always be running ahead `main`, so we will probably pin versions and test them before bumping (on maybe a weekly cadence). If there was a kata-containers cached version of that sha that we could just pick up that would save us building and managing our own cache.

Sorry - maybe off topic and rambling here!

Reply:
Thanks for pointing me to this PR @stevenhorsman !

@fidencio peer pods currently rebuild the kata-agent and other dependencies (attestation-agent, for example). It consumes a significant amount of time that's we are looking forward ways to reduce. One approach is to implement a cache mechanism like in kata; another would be to fetch already built dependencies...you know what is easier :)

Reply:
continuing... kata will be caching the osbuilder entirely, meaning that peer pods will need to open the image to extract the `kata-agent` binary. Is there any problem if we try to cache kata-agent alone?